### ISSUE TITLE
geom_imshow: nice to be able to use color scales.

### ISSUE THREAD

**Issue by alshan** [JetBrains/lets-plot#780]
See feedback: https://github.com/JetBrains/lets-plot/discussions/611#discussioncomment-5907752

> it would be awesome to be able to use the scale_color_gradient... family in `cmap=`

**Comment by egayer** [comment]
Hi @alshan !
thanks, I hope you're well too. For some reason I only this now ! I am sorry for the late answer, but of course I have seen the v4.9.0 as soon as it was out (I am kinda heavy user of lets-plot ;)). I absolutely like these 2 features ! and the v4.9.0. (I am actually about to test/use the geom_bracket() for my next paper!)

To be honest, I don't remember if `cmap` was ever something I told you about (but I might have of course :)). Anyway, I absolutely like it ! It was the only missing piece in `geom_imshow()` and I was patiently waiting because I knew it was coming at some point. However, thanks to scale_cmapmpl(), that you did implement super early on, we have been able to use gradient from scientific colormaps packages. And now, with this new feature in `geom_imshow()` lets-plot is not only just a statistical plotting library. 

Thanks again for the tremendous work you guys are doing and for implementing requests from users. This is really cool.

I am really sorry I have not been able to make a tutorial on mapping with rasters and shapefiles like you mentioned in our last exchange, but I have not forgotten ! (I have been caught up trying to finish my own library :) available since the past week)

Oh and since we are talking about lets-plot capabilities, I know you guys are working on the dual axis in plots, that a lot of people have been asking for, is it something we will see in the coming versions ? I have tried little hacks, but was unable to get a solid alignement. And speaking of hacks, I am currently trying to apply a color scale to labels, I know, I know, sounds weird but could be useful :) can you think of a little trick I could try ?

Thanks again for following up with features we discuss, I really appreciate it, and sorry again for the late answer

**Comment by egayer** [comment]
Hi again, 
speaking of which :)... I just came across a "weird" behavior with `geom_imshow()`: when plotting a 'binary' array containing NaNs and pixels with the same value, `geom_imshow()` plots a square filled of the color attributed to `np.nan` (grey as a default).

This mask thing is really useful when we want to plot mask on a grid. First I thought it was due to the new feature `cmap=` that broke it, but it does not look like it of course :))

However, thanks to the new `cmap=` I found a workaround that we can use until you guys decide if it is worth modifying `geom_imshow()`
Here is the trick : given a mask with NaNs and another value (let say 1), we just need to create a new 2D array with random values (np.random.rand(shape of the mask)). Multiplying this new 2D array by the mask, we get a mask made of NaNs and random values. This way `geom_imshow()` behaves the way it should, which is making the NaNs transparent, and plotting the other pixels with gradient. Now, using `cmap=['#000000','#000000']` for example, then the pixels of random values become all black here.
With this trick, we can plot a mask with patches of the color of our choice, on top of another 2D array, with NaNs being transparent.

PS: I used "weird behavior" because it was not the behavior I expected, but maybe you guys designed `geom_imshow()` using only a range values :)

**Comment by alshan** [comment]
Hi, @egayer !

Yes , we are planning dual axis via ggdeck() composite for the next release.
There is some progress in this area; you may want to take a look at this draft of a demo notebook: [ggdeck_dual_axis.html](https://raw.githack.com/JetBrains/lets-plot/master/docs/f-26b/ggdeck_dual_axis.html)

> And speaking of hacks, I am currently trying to apply a color scale to labels, I know, I know, sounds weird but could be useful :) can you think of a little trick I could try ?

This, unfortunately, we do not support. The only hack I can think of is to hide the axis in the `theme()`, and use `geom_hline()` (to mimic the axis line) along with `geom_text()` for the "axis labels".

----
> a "weird" behavior with geom_imshow()

You are right, I can see the issue and can reproduce it with the following code:
```
image_data = np.array([
    [1.0, np.nan, 1.0],
    [np.nan, 1.0, np.nan],
    [1.0, np.nan, 1.0]
])

ggplot() + geom_imshow(image_data)
``` 

The issue lies in the normalization algorithm when the range is collapsed to zero. If I disable normalization (`geom_imshow(image_data, norm=False)`), I can see a checkerboard with black squares.

As a workaround, could you try normalizing the values manually to the range [0..255]?
```
image_data_norm = np.array([
    [127, np.nan, 127],
    [np.nan, 127, np.nan],
    [127, np.nan, 127]
])

ggplot() + geom_imshow(image_data_norm, norm=False)
```

Would that work for your use case?

**Comment by egayer** [comment]
Thank you @alshan for you answer, 

It’s great to hear that dual axes will be supported in the next release, very good news!

Your workaround for coloring the x-axis labels works perfectly. I had considered something similar, but it felt a bit brute-force, so I was wondering if there might be a more “Lets-Plot-native” approach. In any case, this works great.

Regarding `geom_imshow()`, your workaround also works very well and even better than I expected. I didn’t need to normalize the values: simply using `norm=False` on a 2D array like your` image_data`, with `1.0` and `np.nan`, worked perfectly. I haven’t tested it on other arrays, but for the cases where I encountered the problem, it works perfectly.

Now, while playing with `geom_imshow()` on classified data, I ran into the limitation that it doesn’t support regular lets-plot scales, like `scale_manual()`, where values, breaks, and labels can be explicitly defined. At first, I thought `cguide=guide_colorbar()` might help, but I realized that guides don’t handle breaks or labels.
I know that `geom_imshow() `is more of an auxiliary feature (and I really appreciate you guys having implemented it) and doesn’t fully align with the usual way Lets-Plot handles colors and scales, but I wanted to mention this in case it could be considered for a future release.

**Comment by egayer** [comment]
"As always, we would love to see any new publications where Lets-Plot came in handy! :) Please share."

You mean link to actual articles ? I think I sent you a link to my last one (the first using lets-plot), but here it is again in case it helps : https://agupubs.onlinelibrary.wiley.com/doi/10.1029/2024JF007818
Feel free to use it, lets-plot users will recognize geom_imshow(), geom_map(), geom_lollipop(), geom_boxplot() as well as Marginal Plots and facet_grid()/wrap() !

**Comment by alshan** [comment]
> You mean link to actual articles ? I think I sent you a link to my last one (the first using lets-plot), but here it is again in case it helps : https://agupubs.onlinelibrary.wiley.com/doi/10.1029/2024JF007818

Yes, you did share it earlier—thank you! The visuals are excellently done.

> guides don’t handle breaks or labels.

That shouldn't be a problem to support - I've created an issue: #1486
